lib/Dialect/ATen/ReturnEliminationPass.cpp

```diff
@@ -0,0 +1,182 @@
+      erasedOps.insert(op);
+      auto fn = module.lookupSymbol<FuncOp>(callOp.callee());
+      if (fn && fn.use_empty())
+        erasedOps.insert(fn);
+    } else if (isa<AllocOp>(op)) {
+      Value v = op->getResult(0);
+      if (valueMap.count(v)) {
+        v.replaceAllUsesWith(valueMap[v]);
+        erasedOps.insert(op);
+      }
+    } else {
+      // getModule().dump();
+      // op->dump();
+      // llvm_unreachable("unhandled operation type");
+    }
+
+    for (Value v : op->getOperands()) {
+      if (!v.getType().isa<MemRefType>())
+        continue;
+      if (v.isa<BlockArgument>())
+        continue;
+      if (v.getDefiningOp())
+        runOn(v.getDefiningOp());
+    }
+  }
+
+  void runOnOperation() override {
+
+    auto module = getOperation();
+    auto context = module.getContext();
+
+    // check that a function called "graph" exists
+    auto graph = module.lookupSymbol<mlir::FuncOp>("graph");
+    if (!graph) {
+      emitError(mlir::UnknownLoc::get(module.getContext()),
+                "OpReportPass failed: can't find a graph function\n");
+      signalPassFailure();
+      return;
+    }
+
+    // assume a single bb with a single return statement
+    Block &BB = graph.front();
+
+    FunctionType funcTy = graph.getType();
+    std::vector<Type> newFuncInputTys;
+
+    for (auto ty : funcTy.getInputs())
+      newFuncInputTys.push_back(ty);
+
+    for (auto ty : funcTy.getResults())
+      newFuncInputTys.push_back(ty);
+
+    FunctionType newFuncTy =
+        FunctionType::get(newFuncInputTys, {}, module.getContext());
+    graph.setType(newFuncTy);
+
+    Operation *retOp = BB.getTerminator();
+    auto builder = std::make_unique<mlir::OpBuilder>(retOp);
+
+    builder->create<ReturnOp>(retOp->getLoc());
+
+    std::vector<Value> operands{retOp->getOperands().begin(),
+                                retOp->getOperands().end()};
+
+    retOp->dropAllReferences();
+    erasedOps.insert(retOp);
+
+    for (Value v : operands)
+      valueMap[v] = BB.addArgument(v.getType());
+
+    for (Value v : operands) {
+      if (!v.getType().isa<MemRefType>())
+        llvm_unreachable("graph function returns non-memref");
+      if (v.getDefiningOp())
+        runOn(v.getDefiningOp());
+    }
+
+    for (auto oi = BB.rbegin(), oe = BB.rend(); oi != oe; ++oi) {
+      Operation *o = &*oi;
+      for (Value v : o->getResults()) {
+        if (v.getType().isa<MemRefType>()) {
+          runOn(o);
+          break;
+        }
+      }
+    }
+
+    for (Operation *o : erasedOps)
+      o->erase();
+  }
+
+private:
+  llvm::DenseMap<Value, Value> valueMap;
+  std::set<Operation *> visitedOps;
+  std::set<Operation *> erasedOps;
+};
+
+} // namespace
+
+std::unique_ptr<mlir::Pass> mlir::NPCOMP::aten::createReturnEliminationPass() {
+  return std::make_unique<ReturnEliminationPass>();
+}
+
+void mlir::NPCOMP::aten::registerReturnEliminationPass() {
```

Comment:
do we have tests for this pass? I didn't see return-elimination in any tests.

Reply:
hrm.. good point.  It's mostly used in the pytorch integration.

Reply:
moved to issue